src/batch/src/executor/hash_agg.rs

```diff
@@ -260,53 +534,157 @@ impl<K: HashKey + Send + Sync> HashAggExecutor<K> {
                 }
+                let agg_state_row = OwnedRow::from_iter(agg_datums.into_iter());
+                let hash_code = agg_spill_manager.spill_build_hasher.hash_one(key);
+                agg_spill_manager
+                    .write_agg_state_row(key_row.chain(agg_state_row), hash_code)
+                    .await?;
+            }
+
+            // Release memory occupied by agg hash map.
+            self.mem_context.add(memory_usage_diff);
+
+            // Spill input chunks.
+            #[for_await]
+            for chunk in input_stream {
+                let chunk: DataChunk = chunk?;
+                let hash_codes = chunk.get_hash_values(
+                    self.group_key_columns.as_slice(),
+                    agg_spill_manager.spill_build_hasher,
+                );
+                agg_spill_manager
+                    .write_input_chunk(
+                        chunk,
+                        hash_codes
+                            .into_iter()
+                            .map(|hash_code| hash_code.value())
+                            .collect(),
+                    )
+                    .await?;
             }
-            if !has_next {
-                break; // exit loop
+
+            agg_spill_manager.close_writers().await?;
+
+            // Process each partition one by one.
+            for i in 0..agg_spill_manager.partition_num {
+                let agg_state_stream = agg_spill_manager.read_agg_state_partition(i).await?;
+                let input_stream = agg_spill_manager.read_input_partition(i).await?;
+
+                let sub_hash_agg_executor: HashAggExecutor<K> = HashAggExecutor::new(
+                    self.aggs.clone(),
+                    self.group_key_columns.clone(),
+                    self.group_key_types.clone(),
+                    self.schema.clone(),
+                    Box::new(WrapStreamExecutor::new(child_schema.clone(), input_stream)),
+                    Some(Box::new(WrapStreamExecutor::new(
+                        self.schema.clone(),
+                        agg_state_stream,
+                    ))),
+                    format!("{}-sub{}", self.identity.clone(), i),
+                    self.chunk_size,
+                    self.mem_context.clone(),
+                    self.enable_spill,
+                    self.shutdown_rx.clone(),
+                );
+
+                debug!(
+                    "create sub_hash_agg {} for hash_agg {} to spill",
+                    sub_hash_agg_executor.identity, self.identity
+                );
+
+                let sub_hash_agg_stream = Box::new(sub_hash_agg_executor).execute();
+
+                #[for_await]
+                for chunk in sub_hash_agg_stream {
+                    let chunk = chunk?;
+                    yield chunk;
+                }
+
+                // Clear files of the current partition.
+                agg_spill_manager.clear_partition(i).await?;
```

Comment:
`SpillOp::drop()` also clean files. What's their difference?

Reply:
This can reclaim the storage early.